File: backend/src/bond_monitor/domain/notifications/rules.py
"""Extensible portfolio alert rules."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from datetime import date
from typing import Literal, Protocol

from bond_monitor.domain.bonds.models import BondRecord
from bond_monitor.domain.bonds.offers import (
    bond_offer_view,
    put_offer_action_message,
    put_offer_awareness_message,
)
from bond_monitor.domain.notifications.models import Alert, AlertKind
from bond_monitor.domain.notifications.policies import DEFAULT_NOTIFICATION_POLICY, NotificationPolicy
from bond_monitor.domain.portfolio.models import Portfolio, PortfolioPosition
from bond_monitor.domain.portfolio.policies import DEFAULT_RISK_MONITOR_POLICY, RiskMonitorPolicy
from bond_monitor.domain.portfolio.put_offer import put_offer_awareness_due, put_offer_submit_due
from bond_monitor.domain.portfolio.risk_monitor import detect_risk_escalations, risk_snapshot_from_bond
from bond_monitor.domain.trading.holdings import HoldingView
from bond_monitor.domain.trading.policies import (
    reference_market_price_pct,
    sell_limit_price_buffer,
    suggested_sell_limit_price_pct,
)


@dataclass(frozen=True)
class AlertContext:
    portfolio: Portfolio
    holdings: list[HoldingView]
    positions: list[PortfolioPosition]
    universe: Sequence[BondRecord]
    today: date
    notification_policy: NotificationPolicy = DEFAULT_NOTIFICATION_POLICY
    risk_policy: RiskMonitorPolicy = DEFAULT_RISK_MONITOR_POLICY

    @property
    def universe_by_isin(self) -> dict[str, BondRecord]:
        return {bond.isin: bond for bond in self.universe}


class AlertRule(Protocol):
    def evaluate(self, ctx: AlertContext) -> list[Alert]: ...


class PutOfferActionRule:
    def evaluate(self, ctx: AlertContext) -> list[Alert]:
        alerts: list[Alert] = []
        for position in ctx.positions:
            view = bond_offer_view(position, ctx.today)
            if view is None:
                continue
            if not put_offer_submit_due(position, ctx.today):
                continue
            days_until = (
                (view.submission_end - ctx.today).days
                if view.submission_end is not None
                else (view.execution_date - ctx.today).days
            )
            urgency: Literal["normal", "soon", "critical"] = (
                "critical" if days_until <= 7 else "soon"
            )
            detail_key = (
                view.submission_end.isoformat()
                if view.submission_end is not None
                else view.execution_date.isoformat()
            )
            template = (
                f"Здравствуйте! Прошу принять к исполнению заявку на досрочное погашение "
                f"облигаций {position.name} (ISIN {position.isin}) по пут-оферте."
            )
            alerts.append(
                Alert(
                    portfolio_id=ctx.portfolio.id,
                    kind=AlertKind.PUT_OFFER_ACTION,
                    isin=position.isin,
                    name=position.name,
                    lots=position.lots,
                    figi=position.figi,
                    reason=put_offer_action_message(view),
                    urgency=urgency,
                    detail_key=detail_key,
                    due_date=view.submission_end or view.execution_date,
                    chat_template=template,
                    suggested_price_pct=position.offer_price_pct,
                    offer_window_status=view.window_status.value,
                    submission_start=view.submission_start,
                    submission_end=view.submission_end,
                )
            )
        return alerts


class PutOfferWatchRule:
    """Awareness alerts — used by trading advise, not outbound worker by default."""

    def evaluate(self, ctx: AlertContext) -> list[Alert]:
        if not ctx.notification_policy.include_put_offer_watch_in_alerts:
            return []
        alerts: list[Alert] = []
        for position in ctx.positions:
            view = bond_offer_view(position, ctx.today)
            if view is None or not put_offer_awareness_due(position, ctx.today):
                continue
            if put_offer_submit_due(position, ctx.today):
                continue
            alerts.append(
                Alert(
                    portfolio_id=ctx.portfolio.id,
                    kind=AlertKind.PUT_OFFER_WATCH,
                    isin=position.isin,
                    name=position.name,
                    lots=position.lots,
                    figi=position.figi,
                    reason=put_offer_awareness_message(view),
                    urgency="normal",
                    detail_key=view.execution_date.isoformat(),
                    due_date=view.execution_date,
                    offer_window_status=view.window_status.value,
                    submission_start=view.submission_start,
                    submission_end=view.submission_end,
                    suggested_price_pct=position.offer_price_pct,
                )
            )
        return alerts


class RiskEscalationRule:
    def evaluate(self, ctx: AlertContext) -> list[Alert]:
        alerts: list[Alert] = []
        for holding in ctx.holdings:
            if not holding.isin or holding.lots <= 0:
                continue
            baseline = ctx.portfolio.risk_baselines.get(holding.isin)
            if baseline is None:
                continue
            bond = ctx.universe_by_isin.get(holding.isin)
            if bond is None:
                continue
            current = risk_snapshot_from_bond(bond)
            escalations = detect_risk_escalations(
                baseline,
                current,
                policy=ctx.risk_policy,
            )
            if not escalations:
                continue
            urgency: Literal["normal", "soon", "critical"] = (
                "critical" if any(e.urgency == "critical" for e in escalations) else "soon"
            )
            reason = "; ".join(e.reason for e in escalations)
            market_price = reference_market_price_pct(
                bond_last_price=bond.last_price,
                broker_current_price_pct=holding.current_price_pct,
            )
            buffer = sell_limit_price_buffer(ctx.portfolio.account_kind)
            suggested_price = float(suggested_sell_limit_price_pct(market_price, buffer))
            primary = escalations[0]
            alerts.append(
                Alert(
                    portfolio_id=ctx.portfolio.id,
                    kind=AlertKind.RISK_ESCALATION,
                    isin=holding.isin,
                    name=holding.name,
                    lots=holding.lots,
                    figi=holding.figi,
                    reason=f"Ухудшение риск-профиля эмитента: {reason}. Рекомендуем продать.",
                    urgency=urgency,
                    detail_key=primary.kind,
                    suggested_price_pct=suggested_price,
                    market_price_pct=market_price,
                    risk_acknowledgeable=True,
                    escalation_kinds=tuple(e.kind for e in escalations),
                )
            )
        return alerts


DEFAULT_ALERT_RULES: list[AlertRule] = [
    PutOfferActionRule(),
    PutOfferWatchRule(),
    RiskEscalationRule(),
]

WORKER_ALERT_RULES: list[AlertRule] = [
    PutOfferActionRule(),
    RiskEscalationRule(),
]


def collect_alerts(
    portfolio: Portfolio,
    *,
    holdings: list[HoldingView],
    positions: list[PortfolioPosition],
    universe: Sequence[BondRecord],
    today: date,
    rules: Sequence[AlertRule] | None = None,
    notification_policy: NotificationPolicy = DEFAULT_NOTIFICATION_POLICY,
    risk_policy: RiskMonitorPolicy = DEFAULT_RISK_MONITOR_POLICY,
) -> list[Alert]:
    ctx = AlertContext(
        portfolio=portfolio,
        holdings=holdings,
        positions=positions,
        universe=universe,
        today=today,
        notification_policy=notification_policy,
        risk_policy=risk_policy,
    )
    active_rules = list(rules) if rules is not None else list(DEFAULT_ALERT_RULES)
    alerts: list[Alert] = []
    for rule in active_rules:
        alerts.extend(rule.evaluate(ctx))
    return alerts

File: backend/internal/domain/notifications/rules.go
package notifications

import (
	"fmt"
	"strings"
	"time"

	"github.com/tonatos/bond-monitor/backend/internal/domain/bonds"
	"github.com/tonatos/bond-monitor/backend/internal/domain/market_signals"
	"github.com/tonatos/bond-monitor/backend/internal/domain/portfolio"
	"github.com/tonatos/bond-monitor/backend/internal/domain/shared"
)

// HoldingSnapshot is a minimal holding view for alert rules.
type HoldingSnapshot struct {
	ISIN              string
	FIGI              string
	Name              string
	Lots              int
	CurrentPricePct   *float64
}

// AlertContext carries inputs for alert rule evaluation.
type AlertContext struct {
	Portfolio           portfolio.Portfolio
	Holdings            []HoldingSnapshot
	Positions           []portfolio.PortfolioPosition
	Universe            []bonds.BondRecord
	Today               time.Time
	KeyRatePP           float64
	TaxRateFraction     float64
	NotificationPolicy  NotificationPolicy
	RiskPolicy          portfolio.RiskMonitorPolicy
}

func (ctx AlertContext) universeByISIN() map[string]bonds.BondRecord {
	m := make(map[string]bonds.BondRecord, len(ctx.Universe))
	for _, bond := range ctx.Universe {
		m[bond.ISIN] = bond
	}
	return m
}

// AlertRule evaluates one class of portfolio alerts.
type AlertRule interface {
	Evaluate(ctx AlertContext) []Alert
}

type PutOfferActionRule struct{}

func (PutOfferActionRule) Evaluate(ctx AlertContext) []Alert {
	var alerts []Alert
	for _, position := range ctx.Positions {
		view := bonds.BondOfferViewFrom(position, ctx.Today)
		if view == nil || !portfolio.PutOfferSubmitDue(position, ctx.Today) {
			continue
		}
		daysUntil := shared.DaysBetween(ctx.Today, view.ExecutionDate)
		if view.SubmissionEnd != nil {
			daysUntil = shared.DaysBetween(ctx.Today, *view.SubmissionEnd)
		}
		urgency := AlertUrgencySoon
		if daysUntil <= 7 {
			urgency = AlertUrgencyCritical
		}
		detailKey := shared.FormatISODate(view.ExecutionDate)
		if view.SubmissionEnd != nil {
			detailKey = shared.FormatISODate(*view.SubmissionEnd)
		}
		template := fmt.Sprintf(
			"Здравствуйте! Прошу принять к исполнению заявку на досрочное погашение облигаций %s (ISIN %s) по пут-оферте.",
			position.Name, position.ISIN,
		)
		due := view.ExecutionDate
		if view.SubmissionEnd != nil {
			due = *view.SubmissionEnd
		}
		windowStatus := string(view.WindowStatus)
		alerts = append(alerts, Alert{
			PortfolioID: ctx.Portfolio.ID, Kind: AlertKindPutOfferAction,
			ISIN: position.ISIN, Name: position.Name, Lots: position.Lots, FIGI: position.FIGI,
			Reason: bonds.PutOfferActionMessage(*view), Urgency: urgency, DetailKey: detailKey,
			DueDate: &due, ChatTemplate: &template, SuggestedPricePct: position.OfferPricePct,
			OfferWindowStatus: &windowStatus,
			SubmissionStart: view.SubmissionStart, SubmissionEnd: view.SubmissionEnd,
		})
	}
	return alerts
}

type PutOfferWatchRule struct{}

func (PutOfferWatchRule) Evaluate(ctx AlertContext) []Alert {
	if !ctx.NotificationPolicy.IncludePutOfferWatchInAlerts {
		return nil
	}
	var alerts []Alert
	for _, position := range ctx.Positions {
		view := bonds.BondOfferViewFrom(position, ctx.Today)
		if view == nil || !portfolio.PutOfferAwarenessDue(position, ctx.Today) {
			continue
		}
		if portfolio.PutOfferSubmitDue(position, ctx.Today) {
			continue
		}
		windowStatus := string(view.WindowStatus)
		alerts = append(alerts, Alert{
			PortfolioID: ctx.Portfolio.ID, Kind: AlertKindPutOfferWatch,
			ISIN: position.ISIN, Name: position.Name, Lots: position.Lots, FIGI: position.FIGI,
			Reason: bonds.PutOfferAwarenessMessage(*view), Urgency: AlertUrgencyNormal,
			DetailKey: shared.FormatISODate(view.ExecutionDate), DueDate: &view.ExecutionDate,
			OfferWindowStatus: &windowStatus,
			SubmissionStart: view.SubmissionStart, SubmissionEnd: view.SubmissionEnd,
			SuggestedPricePct: position.OfferPricePct,
		})
	}
	return alerts
}

type RiskEscalationRule struct{}

func (RiskEscalationRule) Evaluate(ctx AlertContext) []Alert {
	universeByISIN := ctx.universeByISIN()
	var alerts []Alert
	for _, holding := range ctx.Holdings {
		if holding.ISIN == "" || holding.Lots <= 0 {
			continue
		}
		baseline, ok := ctx.Portfolio.RiskBaselines[holding.ISIN]
		if !ok {
			continue
		}
		bond, ok := universeByISIN[holding.ISIN]
		if !ok {
			continue
		}
		current := portfolio.RiskSnapshotFromBond(bond)
		escalations := portfolio.DetectRiskEscalations(baseline, current, ctx.RiskPolicy)
		if len(escalations) == 0 {
			continue
		}
		urgency := AlertUrgencySoon
		for _, e := range escalations {
			if e.Urgency == "critical" {
				urgency = AlertUrgencyCritical
				break
			}
		}
		var reasons []string
		var kinds []string
		for _, e := range escalations {
			reasons = append(reasons, e.Reason)
			kinds = append(kinds, string(e.Kind))
		}
		marketPrice := portfolio.ReferenceMarketPricePct(bond.LastPrice, holding.CurrentPricePct, 100)
		buffer := portfolio.SellLimitPriceBuffer(ctx.Portfolio.AccountKind)
		suggested := float64(portfolio.SuggestedSellLimitPricePct(marketPrice, buffer))
		reason := "Ухудшение риск-профиля эмитента: " + strings.Join(reasons, "; ") + ". Рекомендуем продать."
		alerts = append(alerts, Alert{
			PortfolioID: ctx.Portfolio.ID, Kind: AlertKindRiskEscalation,
			ISIN: holding.ISIN, Name: holding.Name, Lots: holding.Lots, FIGI: strPtr(holding.FIGI),
			Reason: reason, Urgency: urgency, DetailKey: kinds[0],
			SuggestedPricePct: &suggested, MarketPricePct: &marketPrice,
			RiskAcknowledgeable: true, EscalationKinds: kinds,
		})
	}
	return alerts
}

type SpreadAnomalyRule struct {
	Policy market_signals.SpreadAnomalyPolicy
}

func (r SpreadAnomalyRule) Evaluate(ctx AlertContext) []Alert {
	policy := r.Policy
	if policy.MinPeers == 0 {
		policy = market_signals.DefaultSpreadAnomalyPolicy
	}
	universeByISIN := ctx.universeByISIN()
	var alerts []Alert

	for _, holding := range ctx.Holdings {
		if holding.ISIN == "" || holding.Lots <= 0 {
			continue
		}
		bond, ok := universeByISIN[holding.ISIN]
		if !ok {
			continue
		}
		if bond.Sector == "" {
			continue
		}
		targetSpread := market_signals.CreditSpreadPP(bond, ctx.KeyRatePP, ctx.TaxRateFraction)
		if targetSpread == nil {
			continue
		}
		peers := market_signals.PeerGroup(bond, ctx.Universe, policy)
		if len(peers) < policy.MinPeers {
			continue
		}
		spreads := make([]float64, 0, len(peers))
		for _, p := range peers {
			if s := market_signals.CreditSpreadPP(p, ctx.KeyRatePP, ctx.TaxRateFraction); s != nil {
				spreads = append(spreads, *s)
			}
		}
		stats := market_signals.SpreadStatsFromPeers(spreads)
		if stats == nil || stats.Peers < policy.MinPeers {
			continue
		}
		anomaly := *targetSpread - stats.ExpectedPP
		z := market_signals.ZScore(*targetSpread, stats.ExpectedPP, stats.StdDevPP)
		isAnomaly := anomaly >= policy.MinAnomalyPP
		if z != nil && *z >= policy.MinZScore {
			isAnomaly = true
		}
		if !isAnomaly {
			continue
		}
		alerts = append(alerts, Alert{
			PortfolioID: ctx.Portfolio.ID, Kind: AlertKindSpreadAnomaly,
			ISIN: holding.ISIN, Name: holding.Name, Lots: holding.Lots, FIGI: strPtr(holding.FIGI),
			Reason: fmt.Sprintf(
				"Кредитный спред расширился относительно похожих бумаг: %.1f п.п. vs медиана %.1f п.п. (Δ %.1f п.п., peers %d).",
				*targetSpread, stats.ExpectedPP, anomaly, stats.Peers,
			),
			Urgency:   AlertUrgencyNormal,
			DetailKey: bond.Sector,
		})
	}
	return alerts
}

type SectorConcentrationRule struct {
	MaxSectorShare float64
}

func (r SectorConcentrationRule) Evaluate(ctx AlertContext) []Alert {
	universeByISIN := ctx.universeByISIN()
	maxShare := r.MaxSectorShare
	if maxShare <= 0 {
		maxShare = portfolio.DefaultDiversificationPolicy.MaxSectorShare
	}

	lotsByISIN := make(map[string]int)
	if len(ctx.Holdings) > 0 {
		for _, h := range ctx.Holdings {
			if h.ISIN != "" && h.Lots > 0 {
				lotsByISIN[h.ISIN] += h.Lots
			}
		}
	} else {
		for _, p := range ctx.Positions {
			if p.ISIN != "" && p.Lots > 0 {
				lotsByISIN[p.ISIN] += p.Lots
			}
		}
	}

	totalValue := 0.0
	for isin, lots := range lotsByISIN {
		b, ok := universeByISIN[isin]
		if !ok {
			continue
		}
		if p := b.PricePerLotRub(); p != nil && *p > 0 {
			totalValue += *p * float64(lots)
		}
	}
	if totalValue <= 0 {
		return nil
	}
	exposures := portfolio.ExposureBySector(universeByISIN, lotsByISIN, totalValue)
	for _, e := range exposures {
		if e.Key == "unknown" {
			continue
		}
		if e.Share <= maxShare {
			continue
		}
		sectorLabel := e.Key
		sharePct := e.Share * 100
		alerts := []Alert{{
			PortfolioID: ctx.Portfolio.ID,
			Kind:        AlertKindSectorConcentration,
			ISIN:        "sector:" + e.Key,
			Name:        "Концентрация в секторе: " + sectorLabel,
			Lots:        0,
			Reason: fmt.Sprintf(
				"Сектор «%s» занимает %.1f%% портфеля (лимит %.0f%%). Рекомендуем диверсифицировать.",
				sectorLabel, sharePct, maxShare*100,
			),
			Urgency:   AlertUrgencyNormal,
			DetailKey: e.Key,
		}}
		return alerts
	}
	return nil
}

func strPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

var DefaultAlertRules = []AlertRule{
	PutOfferActionRule{},
	PutOfferWatchRule{},
	RiskEscalationRule{},
	SectorConcentrationRule{},
	SpreadAnomalyRule{},
}

var WorkerAlertRules = []AlertRule{
	PutOfferActionRule{},
	RiskEscalationRule{},
	SectorConcentrationRule{},
	SpreadAnomalyRule{},
}

// AlertParams groups inputs for CollectAlerts.
type AlertParams struct {
	Portfolio          portfolio.Portfolio
	Holdings           []HoldingSnapshot
	Positions          []portfolio.PortfolioPosition
	Universe           []bonds.BondRecord
	Today              time.Time
	KeyRatePP          float64
	TaxRateFraction    float64
	Rules              []AlertRule
	NotificationPolicy NotificationPolicy
	RiskPolicy         portfolio.RiskMonitorPolicy
}

// CollectAlerts runs alert rules and returns detected events.
func CollectAlerts(params AlertParams) []Alert {
	ctx := AlertContext{
		Portfolio: params.Portfolio, Holdings: params.Holdings, Positions: params.Positions,
		Universe: params.Universe, Today: params.Today,
		KeyRatePP: params.KeyRatePP, TaxRateFraction: params.TaxRateFraction,
		NotificationPolicy: params.NotificationPolicy, RiskPolicy: params.RiskPolicy,
	}
	rules := params.Rules
	if len(rules) == 0 {
		rules = DefaultAlertRules
	}
	var alerts []Alert
	for _, rule := range rules {
		alerts = append(alerts, rule.Evaluate(ctx)...)
	}
	return alerts
}
